DataConnectors/Agari/AzureFunctionAgari/run.ps1

```diff
@@ -0,0 +1,450 @@
+$client_secret = $env:clientSecret
+
+# Define the Log Analytics Workspace ID and Keys
+$CustomerId = $env:workspaceId
+$SharedKey = $env:workspaceKey
+
+# Define the Graph API credentials
+$GraphTenantId = $env:GraphTenantId
+$GraphClientId = $env:GraphClientId
+$GraphClientSecret = $env:GraphClientSecret
+
+#Define Product(s) Enabled
+$bpEnabled = $env:enableBrandProtectionAPI
+$aprEnabled = $env:enablePhishingResponseAPI
+$apdEnabled = $env:enablePhishingDefenseAPI
+$sgEnabled = $env:enableSecurityGraphSharing
+
+#Function App Configuration for Timer Functions
+$resGrp = $env:resGroup
+$appName = $env:functionName
+$subid = $env:subId
+$bplastLog = $env:BPlastLogTime
+$apdlastLog = $env:APDlastLogTime
+$aprlastLog = $env:APRlastLogTime
+
+# Function to build the Authorization signature for the Log Analytics Data Connector API
+Function Build-Signature ($customerId, $sharedKey, $date, $contentLength, $method, $contentType, $resource)
+{
+    $xHeaders = "x-ms-date:" + $date
+    $stringToHash = $method + "`n" + $contentLength + "`n" + $contentType + "`n" + $xHeaders + "`n" + $resource
+
+    $bytesToHash = [Text.Encoding]::UTF8.GetBytes($stringToHash)
+    $keyBytes = [Convert]::FromBase64String($sharedKey)
+
+    $sha256 = New-Object System.Security.Cryptography.HMACSHA256
+    $sha256.Key = $keyBytes
+    $calculatedHash = $sha256.ComputeHash($bytesToHash)
+    $encodedHash = [Convert]::ToBase64String($calculatedHash)
+    $authorization = 'SharedKey {0}:{1}' -f $customerId,$encodedHash
+    
+    # Dispose SHA256 from heap before return.
+    $sha256.Dispose()
+
+    return $authorization
+}
+
+# Function to create and invoke an API POST request to the Log Analytics Data Connector API
+Function Post-LogAnalyticsData($customerId, $sharedKey, $body, $logType)
+{
+    $method = "POST"
+    $contentType = "application/json"
+    $resource = "/api/logs"
+    $rfc1123date = [DateTime]::UtcNow.ToString("r")
+    $contentLength = $body.Length
+    $signature = Build-Signature `
+        -customerId $customerId `
+        -sharedKey $sharedKey `
+        -date $rfc1123date `
+        -contentLength $contentLength `
+        -method $method `
+        -contentType $contentType `
+        -resource $resource
+    $uri = "https://" + $customerId + ".ods.opinsights.azure.com" + $resource + "?api-version=2016-04-01"
+
+    $headers = @{
+        "Authorization" = $signature;
+        "Log-Type" = $logType;
+        "x-ms-date" = $rfc1123date;
+        "time-generated-field" = $TimeStampField;
+    }
+
+    $response = Invoke-WebRequest -Uri $uri -Method $method -ContentType $contentType -Headers $headers -Body $body -UseBasicParsing
+    return $response.StatusCode
+}
+
+# Timer Function to stamp the last successful log into the environment variable
+Function SetLastLogTime ($subid,$resGrp,$appName,$nextStartDate,$apdLogSuccess,$aprLogSuccess,$bpLogSuccess)
+{
+    #Get the subscription
+    Select-AzSubscription -Subscriptionid $subid
```

Comment:
I followed the instruction Option2 to install the Azure function manually.  And the invocation of the function failed with the following error, please make sure this is fully tested:

This line throws exception: ERROR: No subscription found in the context. Please ensure that the credentials you provided are authorized to access an Azure subscription, then run Connect-AzAccount to login. Exception : Type : System.ApplicationException TargetSite : Name : CreateArmClient DeclaringType : Microsoft.Azure.Commands.Common.Authentication.Factories.ClientFactory MemberType : Method Module : Microsoft.Azure.PowerShell.Authentication.dll StackTrace : at Microsoft.Azure.Commands.Common.Authentication.Factories.ClientFactory.CreateArmClient[TClient](IAzureContext context, String endpoint) at Microsoft.Azure.Commands.WebApps.Utilities.WebsitesClient..ctor(IAzureContext context) at

Reply:
The manual instructions are missing one step, which is to enable the system assigned Identity. Auto-deploy enables it automatically. will include in next commit.

Reply:
It seems the issue has not been resolved by the latest change

Reply:
thanks @nazang  - it seems that the apps needs to have a subscription level role to modify the environment variable, and I added that to the instructions (both for auto-deploy and manual).  I'll look into the file method if there is pushback from the customers.

Reply:
Thank you for the fix